#27 Sensors draft

Munroe-Parlee opened the pull request:
Adds (non BMS) sensors functions for:
-Pin abstraction
-Sensor polling
-Struct containing sensor values
-Operations for writing sensor data to flash and retrieving sensor data from flash

Notes for debug:
-Placeholder sensor storage sector
-Placeholder sensor return values, awaiting parts survey
-Placeholder adc pin handlers
-Intend to rewrite Flash functions to save/ retrieve as an array of values

Munroe-Parlee committed:
obc init and get functions defined

Munroe-Parlee committed:
obc init and get functions defined-forgot to save

Munroe-Parlee committed:
declared sensor read value functions, defined get_sensors_data function

Munroe-Parlee committed:
documented obc.c, obc.h

Munroe-Parlee committed:
Major restructure: made hardware abstraction wrapper functions to model interacting with the sensor/flash memory environment, moved functions from obc to obc_interface Inc and Src files, structured program to read and write sensor data from/to a dedicated flash memory. p.s.: mb gang, should have committed this incrementally :skull:

Munroe-Parlee committed:
defined get acceleration and gyroscope axis get functions with dummy values for compilation reasons, renamed sensor data update function from get_sensor_data to read_sensors in order to match ethans function call in branch 33-obc-main Src->obc.c

Munroe-Parlee committed:
added blink led functions and changed dummy values for testing purposes

Munroe-Parlee committed:
added a function to test value return

Munroe-Parlee committed:
defined read_sensor_flash_data in obc_interface.h to return data from memory

Munroe-Parlee committed:
partial restore of wiped code from documentation

Munroe-Parlee committed:
added includes, define flash save location

Munroe-Parlee committed:
removed non documented functions, updated function names

Munroe-Parlee committed:
refilled save_sensor_data_to_flash and load_sensor_data_from_flash from bms example

Munroe-Parlee committed:
saving changes

Munroe-Parlee committed:
added ADC handlers for the temperature and pressure sensors

Munroe-Parlee committed:
wrote draft pin handler code for temperature, pressure ADC

Munroe-Parlee committed:
Merged dev branch changes into feature branch

Manually de-merge conflicted obc_interface Inc and Src files.
Accepted dev version of all other files.

Munroe-Parlee committed:
init_sensors typo fixed

Munroe-Parlee committed:
marked the start and end of sensor code sections (adc handlers and functions)

zvdv commented on project/Core/Src/obc_interface.c:
```diff
@@ -126,6 +131,114 @@ void load_battery_data_from_flash() {
+    uint32_t pageError;
+
+    erase.TypeErase = FLASH_TYPEERASE_SECTORS;       // Erase by sector
+    erase.Sector = FLASH_SECTOR_7;                   // Make sure this is correct for your chip!
```
I don't know much about handling flash storage, but is this sector meant to be the same as the FLASH_SENSOR_ADDRESS sector?

AniMB replied:
dont erase sector 7, leland is using it

AniMB replied:
use will be using flash sector 0

Munroe-Parlee replied:
Typo, will update

zvdv reviewed (commented):
Looks good, other than what you already mentioned about replacing placeholders once we know what they should be, and writing the accelerometer and gyroscope read functions. I left a couple comments in obc_interface.c.

AniMB reviewed (changes requested):
You have pushed too many files. Only push those files you have edited. That is only the two files, obc_interface.h and .c files.

Munroe-Parlee commented on project/Core/Src/obc_interface.c:
```diff
@@ -126,6 +131,114 @@ void load_battery_data_from_flash() {
+    } else {
+        memset(&sensor_backup, 0, sizeof(SensorsData));
+    }
+}
```
as written it is intended to write to an object held elsewhere through the sensor_backup pointer, I would like to change that to a direct return during testing but have left it paralleling Lelands code for the time being.

zvdv replied:
If you're not returning anything for now, would be best to mark the function void, or return a dummy variable.

Munroe-Parlee committed:
Update flash sector erase typo

Replaced an incorrect flash sector artifact with the  flash sensor save address variable

Munroe-Parlee commented:
> > > You have pushed too many files. Only push those files you have edited. That is only the two files, obc_interface.h and .c files.
> > 
> > 
> > What would be the most straightforward way to change that?
> 
> There's probably a technically better way, but I think what I did when I had the same issue for my branch was copy my two relevant files to somewhere else, then pull dev into my branch in some way that it overwrites everything (can't remember rn what exactly I did), and then copy back my changes to the two relevant files, then make a new commit.
> 
> In future, I wonder if we should gitignore the whole Debug folder, since it is generated when you build. But then there's also the Drivers and Middlewares folders, which I feel like could also be gitignored, but some people online seem to say they should be tracked, so I'm not sure about those.

That's similar to what I did too, I made a new branch off of dev and checked the two important files into it then PR'd that branch.
